src/p_keyframe.c

```diff
@@ -0,0 +1,646 @@
+    M_FreeArenaCopy(keyframe->activeplats);
+    free(keyframe);
+}
+
+typedef struct elem_s
+{
+    keyframe_t *keyframe;
+    struct elem_s *next;
+    struct elem_s *prev;
+} elem_t;
+
+typedef struct
+{
+    elem_t* top;
+    elem_t* tail;
+    int count;
+} queue_t;
+
+static queue_t queue;
+
+static boolean IsEmpty(void)
+{
+    return queue.top == NULL;
+}
+
+// Add an element to the top of the queue
+static void Push(keyframe_t *keyframe)
+{
+    // Remove oldest element if queue is full
+    if (queue.count == rewind_depth)
+    {
+        elem_t *oldtail = queue.tail;
+        if (oldtail)
+        {
+            queue.tail = oldtail->prev;
+            if (queue.tail)
+            {
+                queue.tail->next = NULL;
+            }
+            else
+            {
+                // Queue becomes empty after removal
+                queue.top = NULL;
+            }
+            FreeKeyframe(oldtail->keyframe);
+            free(oldtail);
+            --queue.count;
+        }
+    }
+
+    elem_t *newelem = calloc(1, sizeof(*newelem));    
+    newelem->keyframe = keyframe;
+    
+    if (IsEmpty())
+    {
+        queue.top = newelem;
+        queue.tail = newelem;
+    }
+    else
+    {
+        newelem->next = queue.top;
+        queue.top->prev = newelem;
+        queue.top = newelem;
+    }
+    ++queue.count;
+}
+
+// Remove and return element from top of queue
+static keyframe_t *Pop(void)
+{
+    if (IsEmpty())
+    {
+        return NULL;
+    }
+    
+    elem_t* temp = queue.top;
+    keyframe_t *keyframe = temp->keyframe;
+    queue.top = temp->next;
+    if (queue.top)
+    {
+        queue.top->prev = NULL;
+    }
+    else
+    {
+        queue.tail = NULL;
+    }
+    free(temp);
+    --queue.count;
+
+    return keyframe;
+}
+
+void P_FreeKeyframeQueue(void)
+{
+    elem_t* current = queue.top;
+    while (current)
+    {
+        elem_t* temp = current;
+        current = current->next;
+        FreeKeyframe(temp->keyframe);
+        free(temp);
+    }
+    memset(&queue, 0, sizeof(queue));
+}
+
+void P_SaveKeyframe(void)
+{
+    int interval_tics = TICRATE * rewind_interval / 1000;
+
+    if (keyframe_tic % interval_tics == 0)
+    {
+        Push(SaveKeyframe());
+    }
+
+    ++keyframe_tic;
+}
+
+void P_LoadKeyframe(void)
+{
+    int interval_tics = TICRATE * rewind_interval / 1000;
+
+    while (1)
+    {
+        keyframe_t *keyframe = Pop();
+        
+        if (!keyframe)
+        {
+            break;
+        }
+
+        if (keyframe->tic > 0 && keyframe_tic - keyframe->tic < interval_tics)
+        {
+            FreeKeyframe(keyframe);
+            continue;
+        }
+
+        LoadKeyframe(keyframe);
+
+        if (!queue.count) // don't delete first keyframe
+        {
+            Push(keyframe);
+        }
+        else
+        {
+            FreeKeyframe(keyframe);
+        }
+
+        G_ClearInput();
+        displaymsg("Restore Keyframe %d", queue.count);
```

Comment:
Could we translate "keyframe" to "level time" or "game time" probably?

Reply:
"Keyframe" is a term used in DSDA-Doom, so users are familiar with it. Nugget uses it too.
We could rename it to "save state", like in emulators.

Reply:
My point is that it reports a keyframe number, which is very abstract, instead of reporting how far into the game I am now. I'd prefer something like "Restore game state at 1:05.39". But I understand if you don't want this because other ports already do it differently.

Reply:
In fact, DSDA-Doom only reports "Restored key frame" and doesn't include a number. You're right; the number isn't helpful. I don't like the time either. Maybe we shouldn't print anything.